orchagent/nhgorch.cpp

```diff
@@ -0,0 +1,1386 @@
+ *
+ * Params:      IN  consumer - The cosumer object.
+ *
+ * Returns:     Nothing.
+ */
+void NhgOrch::doTask(Consumer& consumer)
+{
+    SWSS_LOG_ENTER();
+
+    if (!gPortsOrch->allPortsReady())
+    {
+        return;
+    }
+
+    auto it = consumer.m_toSync.begin();
+
+    while (it != consumer.m_toSync.end())
+    {
+        KeyOpFieldsValuesTuple t = it->second;
+
+        string index = kfvKey(t);
+        string op = kfvOp(t);
+
+        SWSS_LOG_INFO("Next hop group table key %s, op %s",
+                        index.c_str(), op.c_str());
+
+        bool success;
+        const auto& nhg_it = m_syncdNextHopGroups.find(index);
+
+        if (op == SET_COMMAND)
+        {
+            string ips;
+            string aliases;
+            string weights;
+            string mpls_nhs;
+
+            /* Get group's next hop IPs and aliases */
+            for (auto i : kfvFieldsValues(t))
+            {
+                if (fvField(i) == "nexthop")
+                    ips = fvValue(i);
+
+                if (fvField(i) == "ifname")
+                    aliases = fvValue(i);
+
+                if (fvField(i) == "weight")
+                    weights = fvValue(i);
+
+                if (fvField(i) == "mpls_nh")
+                    mpls_nhs = fvValue(i);
+            }
+
+            /* Split ips and alaises strings into vectors of tokens. */
+            vector<string> ipv = tokenize(ips, ',');
+            vector<string> alsv = tokenize(aliases, ',');
+            vector<string> mpls_nhv = tokenize(mpls_nhs, ',');
+
+            /* Create the next hop group key. */
+            string nhg_str;
+
+            for (uint32_t i = 0; i < ipv.size(); i++)
+            {
+                if (i) nhg_str += NHG_DELIMITER;
+                if (!mpls_nhv.empty() && mpls_nhv[i] != "na")
+                {
+                    nhg_str += mpls_nhv[i] + LABELSTACK_DELIMITER;
+                }
+                nhg_str += ipv[i] + NH_DELIMITER + alsv[i];
+            }
+
+            NextHopGroupKey nhg_key = NextHopGroupKey(nhg_str, weights);
+
+            /* If the group does not exist, create one. */
+            if (nhg_it == m_syncdNextHopGroups.end())
+            {
+                SWSS_LOG_NOTICE("Adding next hop group %s with %s",
+                                index.c_str(),
+                                nhg_str.c_str());
+                /*
+                * If we'd have to create a SAI object for the group, and we
+                * already reached the limit, we're going to create a temporary
+                * group, represented by one of it's NH only until we have
+                * enough resources to sync the whole group.  The item is going
+                * to be kept in the sync list so we keep trying to create the
+                * actual group when there are enough resources.
+                */
+                if ((nhg_key.getSize() > 1) &&
+                    (NextHopGroup::getCount() >= m_maxNhgCount))
```

Comment:
you should also get the nexthop groups used by routeorch to compare with maxNhgCount, and vice-versa.

Reply:
NhgOrch owns the count of all next hop groups. See the calls in RouteOrch to gNhgOrch->incNhgCount() and gNhgOrch->decNhgCount()